### PR TITLE
[receiver/splunkhec] expose the Splunk HEC codec

### PR DESCRIPTION
**Pull request by atoulme**
**Description:**
Move out of the splunkhec receiver the codec that translates from Splunk HEC JSON to pdata.

**Link to tracking Issue:**
#21152 

**Testing:**
Test coverage is unchanged on the receiver.

**Documentation:**
N/A

### CODE REVIEW

**Comment by atoulme** [comment]
@MovieStoreGuy this one is for you ;)

File: pkg/codec/splunkhec/splunk_to_logdata.go [hunk 1]
```diff
@@ -87,11 +84,11 @@ func splunkHecToLogData(logger *zap.Logger, events []*splunk.Event, resourceCust
 	return ld, nil
 }
 
-// splunkHecRawToLogData transforms raw splunk event into log
-func splunkHecRawToLogData(sc *bufio.Scanner, query url.Values, resourceCustomizer func(pcommon.Resource), config *Config) (plog.Logs, int) {
+// SplunkHecRawToLogData transforms raw splunk event into log
+func SplunkHecRawToLogData(sc *bufio.Scanner, query url.Values, resourceCustomizer func(pcommon.Resource), config *splunk.HecToOtelAttrs) (plog.Logs, int) {
```
Comment (MovieStoreGuy) [inline pkg/codec/splunkhec/splunk_to_logdata.go:88]: This is a naming stutter which goes against the golang style guide. 

```suggestion
// RawToLogData transforms raw splunk event into log
func RawToLogData(sc *bufio.Scanner, query url.Values, resourceCustomizer func(pcommon.Resource), config *splunk.HecToOtelAttrs) (plog.Logs, int)
 ```

**Review by MovieStoreGuy — COMMENTED** [review]
Just a naming stutter but otherwise, seems fine.

File: pkg/codec/splunkhec/splunk_to_logdata.go (hunk 1 above)
Comment (atoulme) [reply pkg/codec/splunkhec/splunk_to_logdata.go:88]: hmm indeed and maybe that doesn't need to be extracted. I'll bring it back for now.

**Comment by MovieStoreGuy** [comment]
Can I leave you to resolve the conflicts @atoulme ?

File: pkg/codec/splunkhec/splunkhec_to_metricdata.go [hunk 2]
```diff
@@ -78,32 +76,28 @@ func splunkHecToMetricsData(logger *zap.Logger, events []*splunk.Event, resource
...
-	return md, numDroppedTimeSeries
+	return md
 }
 
-func convertString(logger *zap.Logger, numDroppedTimeSeries *int, metrics pmetric.MetricSlice, metricName string, pointTimestamp pcommon.Timestamp, s string, attributes pcommon.Map) {
+func convertString(logger *zap.Logger, metrics pmetric.MetricSlice, metricName string, pointTimestamp pcommon.Timestamp, s string, attributes pcommon.Map) {
 	// best effort, cast to string and turn into a number
 	dbl, err := strconv.ParseFloat(s, 64)
 	if err != nil {
-		*numDroppedTimeSeries++
```
Comment (mx-psi) [inline pkg/codec/splunkhec/splunkhec_to_metricdata.go:106]: This and the change in line 99 are just cleanup, right?

**Review by mx-psi — APPROVED** [review]

File: pkg/codec/splunkhec/splunkhec_to_metricdata.go (hunk 2 above)
Comment (atoulme) [reply pkg/codec/splunkhec/splunkhec_to_metricdata.go:106]: Correct. This counter was only used in testing, and didn't really do well with the Marshaler interface. There was no point in keeping it.

File: pkg/codec/splunkhec/go.mod [hunk 3]
```diff
@@ -0,0 +1,57 @@
+module github.com/open-telemetry/opentelemetry-collector-contrib/pkg/codec/splunkhec
+
+go 1.19
+
+require (
+	github.com/json-iterator/go v1.1.12
+	github.com/open-telemetry/opentelemetry-collector-contrib/internal/splunk v0.76.3
```
Comment (MovieStoreGuy) [inline pkg/codec/splunkhec/go.mod:7]: set these to `v0.77.0`

**Review by MovieStoreGuy — APPROVED** [review]

**Comment by MovieStoreGuy** [comment]
Looks like `go.mod` needs to be run again.

**Comment by dmitryax** [comment]
@atoulme @MovieStoreGuy, why do we need this functionality exported in `pkg`? Do we want to offer to use these codecs outside of the repo?

**Comment by MovieStoreGuy** [comment]
> @atoulme @MovieStoreGuy, why do we need this functionality exported in `pkg`? Do we want to offer to use these codecs outside of the repo?

I don't believe anything outside of the project would consume it, the intention was to expose the functionality with the encoding extension they I've only just started.

**Pull request closed without merging**